p2p/libp2p/networksharding/listsSharder.go

```diff
@@ -61,65 +72,113 @@ var _ = leadingZerosCount[255]
+		return nil, fmt.Errorf("%w, maxIntraShardObservers should be at least %d", p2p.ErrInvalidValue, minAllowedObservers)
+	}
+	if maxCrossShardObservers < minAllowedObservers {
+		return nil, fmt.Errorf("%w, maxCrossShardObservers should be at least %d", p2p.ErrInvalidValue, minAllowedObservers)
+	}
+	if maxCrossShardObservers+maxIntraShardObservers == 0 {
+		log.Warn("no connections to observers are possible")
 	}
 
-	return &listsSharder{
-		peerShardResolver: resolver,
-		selfPeerId:        selfPeerId,
-		maxPeerCount:      maxPeerCount,
-		maxIntraShard:     maxIntraShard,
-		maxCrossShard:     maxCrossShard,
-		computeDistance:   computeDistanceByCountingBits,
-	}, nil
+	providedPeers := maxIntraShardValidators + maxCrossShardValidators + maxIntraShardObservers + maxCrossShardObservers
+	if providedPeers+minUnknownPeers > maxPeerCount {
+		return nil, fmt.Errorf("%w, maxValidators + maxObservers should be less than %d", p2p.ErrInvalidValue, maxPeerCount)
+	}
+
+	ls := &listsSharder{
+		peerShardResolver:       resolver,
+		selfPeerId:              selfPeerId,
+		maxPeerCount:            maxPeerCount,
+		computeDistance:         computeDistanceByCountingBits,
+		maxIntraShardValidators: maxIntraShardValidators,
+		maxCrossShardValidators: maxCrossShardValidators,
+		maxIntraShardObservers:  maxIntraShardObservers,
+		maxCrossShardObservers:  maxCrossShardObservers,
+	}
+
+	ls.maxUnknown = maxPeerCount - providedPeers
+
+	return ls, nil
 }
 
 // ComputeEvictionList returns the eviction list
 func (ls *listsSharder) ComputeEvictionList(pidList []peer.ID) []peer.ID {
-	evictionProposed := make([]peer.ID, 0)
-	intraShard, crossShard, unknownShard := ls.splitPeerIds(pidList)
-
-	intraShard, e := evict(intraShard, ls.maxIntraShard)
+	peerDistances := ls.splitPeerIds(pidList)
+
+	existingNumIntraShardValidators := len(peerDistances[intraShardValidators])
+	existingNumIntraShardObservers := len(peerDistances[intraShardObservers])
+	existingNumCrossShardValidators := len(peerDistances[crossShardValidators])
+	existingNumCrossShardObservers := len(peerDistances[crossShardObservers])
+	existingNumUnknown := len(peerDistances[unknown])
+
+	var numIntraShardValidators, numCrossShardValidators int
+	var numIntraShardObservers, numCrossShardObservers int
+	var numUnknown, remaining int
+
+	numIntraShardValidators, remaining = computeUsedAndSpare(existingNumIntraShardValidators, ls.maxIntraShardValidators)
+	numCrossShardValidators, remaining = computeUsedAndSpare(existingNumCrossShardValidators, ls.maxCrossShardValidators+remaining)
+	numIntraShardObservers, remaining = computeUsedAndSpare(existingNumIntraShardObservers, ls.maxIntraShardObservers+remaining)
+	numCrossShardObservers, remaining = computeUsedAndSpare(existingNumCrossShardObservers, ls.maxCrossShardObservers+remaining)
+	numUnknown, _ = computeUsedAndSpare(existingNumUnknown, ls.maxUnknown+remaining)
+
+	evictionProposed := evict(peerDistances[intraShardValidators], numIntraShardValidators)
```

Comment:
why do you evict with numIntraShardValidators. Shouldn't you want to keep the most suitable maxIntraShardValidators instead of numIntraShardValidators?

Same for the others

Reply:
as can be seen on lines L155-L159 I computed the minimum between numIntraShardValidators and maxIntraShardValidators as not to allow having more than max `maxIntraShardValidators` peers. These computed values actually mean: "how many peers should I keep" and is the minimum between what we have and the maximum allowed for th category of peers.